File: windows/PhpNitroDesktop.Protocol/HostPort.cs
namespace PhpNitroDesktop.Protocol;

// The Windows counterpart of ios/Sources/PhpNitroGo/HostPort.swift and
// android/go/src/main/java/com/phpnitro/go/HostPort.kt — a fresh port
// rather than a shared reference, since neither of those lives in a
// platform-agnostic package this project could depend on. Used by
// PhpNitroDesktop.App's own `--connect HOST:PORT` flag, the same
// "IP:PORT, optionally prefixed with a scheme" string `phpx serve`
// prints on every platform.
public static class HostPort
{
    public static (string Host, int Port)? Parse(string input)
    {
        var withoutScheme = input.Trim();
        foreach (var prefix in new[] { "http://", "https://" })
        {
            if (withoutScheme.StartsWith(prefix, System.StringComparison.Ordinal))
            {
                withoutScheme = withoutScheme[prefix.Length..];
                break;
            }
        }
        withoutScheme = withoutScheme.TrimEnd('/');

        var colonIndex = withoutScheme.LastIndexOf(':');
        if (colonIndex <= 0 || colonIndex == withoutScheme.Length - 1)
        {
            return null;
        }

        var host = withoutScheme[..colonIndex];
        var portString = withoutScheme[(colonIndex + 1)..];

        if (!int.TryParse(portString, out var port) || port < 1 || port > 65535)
        {
            return null;
        }

        return (host, port);
    }
}

File: macos/Sources/PhpNitroMacEngine/HostPort.swift
import Foundation

/// The macOS counterpart of `ios/Sources/PhpNitroGo/HostPort.swift` and
/// `android/go/src/main/java/com/phpnitro/go/HostPort.kt` — a fresh,
/// macOS-local copy rather than a dependency on `PhpNitroGo` itself,
/// since that target imports UIKit (its `ConnectViewController`/
/// `ScanViewController`) and pulling it into this macOS package would
/// fail the exact same way `PhpNitroMacEngine` already avoids every
/// other UIKit-only target in `../ios/`'s package (see this package's
/// own `Package.swift` docblock). `PhpNitroMacApp`'s `AppDelegate` uses
/// this for its own `--connect HOST:PORT` flag — the same "IP:PORT,
/// optionally prefixed with a scheme" string `phpx serve` prints on
/// every platform.
public enum HostPort {
    public static func parse(_ input: String) -> (host: String, port: Int)? {
        var withoutScheme = input.trimmingCharacters(in: .whitespacesAndNewlines)
        for prefix in ["http://", "https://"] where withoutScheme.hasPrefix(prefix) {
            withoutScheme.removeFirst(prefix.count)
            break
        }
        while withoutScheme.hasSuffix("/") {
            withoutScheme.removeLast()
        }

        guard let colonIndex = withoutScheme.lastIndex(of: ":"), colonIndex != withoutScheme.startIndex else { return nil }

        let host = String(withoutScheme[withoutScheme.startIndex..<colonIndex])
        let portString = String(withoutScheme[withoutScheme.index(after: colonIndex)...])

        guard !host.isEmpty, !portString.isEmpty, let port = Int(portString), (1...65535).contains(port) else { return nil }

        return (host, port)
    }
}
